Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1417 @@
+# Parsing with Haskell (Part 2): Parsing with Happy
+
+<!--
+* [X] refer to LR(1) article
+* [X] link to its user guide
+-->
+
+This is the second and final part of the Parsing with Haskell series. 
+Haven't read the first part yet? You can find it [here](https://serokell.io/blog/lexing-with-alex).
+
+In this article, we will use [Happy](https://www.haskell.org/happy/) to generate a parser that will consume the token stream from the lexer that Alex has generated for us. 
+
+As with Alex, our goal is to provide enough information to be productive with Happy.
+Hence, we won't cover every single feature of it.
+Please refer to the [Happy User Guide](https://monlih.github.io/happy-docs/) for more information.
+
+This tutorial was written using GHC version 9.0.2, Stack resolver LTS-19.8, Alex version 3.2.7.1, and Happy version 1.20.0.
+
+## Our grammar: MiniML
+
+We discussed our grammar in the previous article, but let's reintroduce it here.
+
+MiniML (read as "minimal") is a small grammar based on [ML](https://en.wikipedia.org/wiki/ML_(programming_language)).
+As the name suggests, it has the minimal syntax needed to get you started with Alex and Happy.
+
+By the end of this tutorial, you will be able to parse programs such as the following:
+
+```sml
+let the_answer : int =
+  let a = 20 in
+  let b = 1 in
+  let c = 2 in
+  a * c + b * c
+
+let main (unit : ()) : () =
+  print ("The answer is: " + the_answer)
+```
+
+Some of the features that this tutorial will cover are:
+
+* Variables, constants, function declarations, function applications, local bindings, and operators.
+* Some simple types, as well as arrow types (functions).
+* Conditional expressions: if-then-else and if-then.
+
+In addition, exercises will guide you through extending the language with patterns, list access, and pattern matching. The first part also introduced block comments (in the main text) and rational numbers and line comments (in exercises).
+
+## Happy vs. Megaparsec
```

Comment:
I think we can cut some stuff from this. 

https://docs.google.com/document/d/1_LYHCUzjDbjngeuQfdyCctAdnADEk60oR_-PrJ_nOFY/edit <- This is what I have on offer.

Reply:
I think I agree it makes sense to cut extraneous stuff from here, as it's not the main focus. I'm not exactly convinced the linked blog post is helpful in comparing recursive descent with LR. From a quick Google search, I found https://www.abubalay.com/blog/2021/12/31/lr-control-flow which gets into more detail about LR (I didn't review it in much detail though, so can't vouch there aren't any horrible mistakes lurking), and there's https://tratt.net/laurie/essays/entries/which_parsing_approach.html which doesn't really get into much detail, but gives a reasonable overview of recursive descent vs LL vs LR (and also throws in some discussion of generalized CFG parsers like GLR, GLL and Earley)

Reply:
I've copy-pasted your article, I think it's better overall, as going into details is not our focus. I also added links to the suggested articles instead.